uniswap/uniswap.py

```diff
@@ -1209,6 +1215,51 @@ def get_weth_address(self) -> ChecksumAddress:
             address = self.router.functions.WETH9().call()
         return address
 
+    def get_raw_price(
+        self, 
+        token_in: AddressLike, 
+        token_out: AddressLike,
+        fee: int = 3000) -> float:
+   
+        if token_in == ETH_ADDRESS:
+            token_in = self.get_weth_address()
+        if token_out == ETH_ADDRESS:
+            token_out = self.get_weth_address()
+
+        if self.version == 2:
+            params = [self.w3.toChecksumAddress(token_in), self.w3.toChecksumAddress(token_out)]
+            pair_token = self.factory_contract.functions.getPair(*params).call()
+            token_in_erc20 = _load_contract_erc20(self.w3, self.w3.toChecksumAddress(token_in))
+            token_in_balance = int(token_in_erc20.functions.balanceOf(self.w3.toChecksumAddress(pair_token)).call())
+            token_in_decimals = self.get_token(token_in).decimals
+            token_in_balance = token_in_balance/(10**token_in_decimals)
+
+            token_out_erc20 = _load_contract_erc20(self.w3, self.w3.toChecksumAddress(token_out))
+            token_out_balance = int(token_out_erc20.functions.balanceOf(self.w3.toChecksumAddress(pair_token)).call())
+            token_out_decimals = self.get_token(token_out).decimals
+            token_out_balance = token_out_balance/(10**token_out_decimals)
+
+            raw_price = token_out_balance/token_in_balance
+        else:
+            params = [self.w3.toChecksumAddress(token_in), self.w3.toChecksumAddress(token_out), fee]
+            pool_address = self.factory_contract.functions.getPool(*params).call()
+            pool_contract = _load_contract(
+                    self.w3, abi_name="uniswap-v3/pool", address=pool_address
+                )
+            t0 = pool_contract.functions.token0().call()
+            t1 = pool_contract.functions.token1().call()
+            if t1.lower() == token_in.lower():
+                den0 = self.get_token(token_in).decimals
+                den1 = self.get_token(token_out).decimals
+            else:
+                den0 = self.get_token(token_out).decimals
+                den1 = self.get_token(token_in).decimals
+            sqrtPriceX96 = pool_contract.functions.slot0().call()[0]
+            raw_price = (sqrtPriceX96*sqrtPriceX96*10**den1>>(96*2))/(10**den0)
```

Comment:
Wow, awesome work figuring this out! (I've been meaning to...)

Reply:
It's a kind of magic. Long story short
https://ethereum.stackexchange.com/questions/98685/computing-the-uniswap-v3-pair-price-from-q64-96-number
But I have to notice that mentioned snippet works correctly only if both tokens have `decimals` == 18 that is wrong in the general case. It needs to determine actual values. Also, if we need this price for ETH-USDC pair but `pool.token0`==USDC and `pool.token1`==ETH then we need to use denominators in reverse order. That's why it requires this ugly if-else just before the calculation.
Also, the final impact value differs a bit from uniswap's one. Delta is ~0.0x% I suppose they apply the taker's fee somewhere. I don't need absolute precision in my tasks (and I really doubt someone needs it in this particular case) so I've just omitted that fact in my own realisation.
Regarding other comments - sure, I'll make a fix.